🚀 4단계 - EntityEntry

SightStudio opened the pull request:
리뷰어님 안녕하세요
4단계 - EntityEntry 관련해서 구현했습니다. 

상태중에 READ_ONLY를 실제 트랜잭션을 사용하지않고 어떻게 
구현해야할까 고민하다가 일단 제외하고 나머지 상태에 대해서 테스트코드로 구현하였습니다.

마지막까지 잘부탁드립니다. :)

SightStudio committed:
feat. 요구사항 구현
더티체킹 구현 (비트셋 기반으로)

SightStudio committed:
refactor. 리뷰반영
테스트 클래스에 퍼블릭 제거

SightStudio committed:
refactor. 리뷰반영
DefaultEntityLoaderTest 테스트에서 EntityManger 대신 EntityPersister로 전환

SightStudio committed:
test. 더티체킹에 사용된 쿼리빌더 옵션 테스트 추가

SightStudio committed:
test. DirtyCheckMarker 테스트 추가

SightStudio committed:
refactor. 리뷰반영. snapshot과 비교하여 더티체크를 해야하면 true로 반환하도록 수정

SightStudio committed:
refactor. EntityPersister를 받기 때문에 IdHolder만 넘기도록 수정

SightStudio committed:
refactor. 리뷰반영. 테스트코드 명제 수정

SightStudio committed:
fix. update에서 @DynamicUpdate 제거

SightStudio committed:
fix. 리뷰반영 - hibernate merge 동작 방식 개선

SightStudio committed:
refactor. 리뷰반영 persistContext 부분 테스트 추가

SightStudio committed:
Merge branch 'step3' into step4

# Conflicts:
#	src/main/java/orm/StatefulPersistenceContext.java

SightStudio committed:
refactor. 리뷰반영 fail-fast

SightStudio committed:
refactor. DirtyCheckMarker 생성자 추가

SightStudio committed:
refactor. 리뷰반영. DirtyCheckMarkerTest 실패테스트 추가

SightStudio committed:
refactor. 리뷰반영. merge 부분 earlyReturn

SightStudio committed:
docs. 요구사항 정리

SightStudio committed:
feat. 요구 사항 1 - 엔터티의 상태를 추가를 EntityEntry, EntityEntryContext 추가

SightStudio committed:
feat. 요구 사항 1 - READ_ONLY를 제외한 모든 구현 추가

SightStudio committed:
feat. 요구 사항 1 - READ_ONLY를 제외한 상태 구현 추가

SightStudio committed:
Merge remote-tracking branch 'origin/step4' into step4

# Conflicts:
#	src/main/java/orm/SessionImpl.java

SightStudio committed:
docs. 독스 수정

SightStudio committed:
polish. 미사용 클래스 제거

SightStudio committed:
refactor. 테스트 문구 변경

jinyoungchoi95 commented on src/main/java/orm/life_cycle/SimpleEntityEntry.java:
```diff
@@ -0,0 +1,43 @@
+
+public class SimpleEntityEntry implements EntityEntry {
+
+    private final Object id;
```
`EntityKey`가 id를 가지고 있는데 `EntityEntry`가 id를 가질 필요가 있을까요?

SightStudio replied:
EntityEntry를 가지고 있는게 좀 더 낫겠네요

jinyoungchoi95 commented on src/main/java/orm/life_cycle/SimpleEntityEntry.java:
```diff
@@ -0,0 +1,43 @@
+        this.status = status;
+    }
+
+    public SimpleEntityEntry(Object idValue) {
+        this.id = idValue;
+        this.status = Status.MANAGED;
+    }
```
초기값이 `MANAGED`인 것은 의문인데요. 처음부터 관리되고 있다라고 보기는 어렵지 않나요?

SightStudio replied:
맞습니다. 그래서 사용하지 않았는데 코드가 남아있었네요 지우도록 하겠습니다.

jinyoungchoi95 commented on src/main/java/orm/life_cycle/Status.java:
```diff
@@ -0,0 +1,10 @@
+	READ_ONLY,  // 읽기전용 상태
+	DELETED,    // 삭제 예정 상태
+	GONE,		// 이미 DB에서 삭제된 상태
+	LOADING,    // 지연 로딩 중인 상태
```
`LOADING`이 지연로딩중인 상태는 아니에요. 지연로딩이라고 확정짓는 것이 아닌 영속성 컨텍스트에 영속화되기 전의 상태에 해당합니다.

SightStudio replied:
요거 개발 초기에 적어놨던건데 제가 안뺐네요 주석들 지워놓도록 하겠습니다 :)

jinyoungchoi95 commented on src/main/java/orm/life_cycle/TrackableEntityEntryContext.java:
```diff
@@ -0,0 +1,52 @@
+
+/**
+ * EntityEntryContext에서 엔티티의 변경 이력을 관리하기 위한 클래스
+ * EntityEntryContext를 상속받아 구현하였고, 주로 테스트를 위한 용도로만 사용된다.
```
테스트를 위한 클래스는 main 모듈에서 모두 제거해주세요 :)

SightStudio replied:
이건 테스트모듈로 따로 빼도록 하겠습니다

jinyoungchoi95 commented on src/main/java/orm/SessionImpl.java:
```diff
@@ -11,31 +12,50 @@ public class SessionImpl implements EntityManager {
+        // 엔티티가 이미 1차 캐시에 존재하는 경우
         T entityInContext = persistenceContext.getEntity(clazz, id);
         if (entityInContext != null) {
+            persistenceContext.addEntry(entityInContext, Status.MANAGED);
```
`MANAGED` 상태는 영속화상태에 있는 entity의 상태에 해당합니다. 이미 `PersistenceContext`에 있다면 굳이 entry를 더 넣어줄 필요가 있을까요?

SightStudio replied:
음.. 넵 그렇네요 안넣어도 될것같아요

jinyoungchoi95 commented on src/main/java/orm/SessionImpl.java:
```diff
@@ -11,31 +12,50 @@ public class SessionImpl implements EntityManager {
             return entity;
         }
 
+        // 엔티티가 DB에서 조회해도 없는 경우
+        persistenceContext.removeEntry(entityKey);
```
제거 로직이 있을 필요가 있을까요?

1. entityLoader에서 값을 가져온다.
2. 값이 없으면 null을 반환한다.
3. 영속화 전 `LOADING` 상태로 변환한다.
4. 영속화한다.

SightStudio replied:
원래는 엔티티가 DB에서 조회해도 없는 경우에는 null을 반환하고 엔트리 상태에서도 제거하고 싶었습니다.
흠.. 생각해보니 그냥 LOADING으로 납둬도 될것같네요

jinyoungchoi95 commented on src/main/java/orm/StatefulPersistenceContext.java:
```diff
@@ -23,12 +39,17 @@ public <T> T getEntity(Class<T> clazz, Object id) {
     public <T> T addEntity(T entity) {
-        var entityKey = new EntityKey(new EntityIdHolder<>(entity));
+        var entityKey = EntityKey.ofEntity(entity);
+        if (entityKey.hasNullIdValue()) {
+            return entity;
+        }
```
해당 동작의 의도가 어떻게될까요? id가 null인 entity가 snapshot에 안들어가면 이후 동작이 잘 안그려지네요.
jpa의 동작구조를 생각해보면 id가 null인 것이 PersistenceContext까지 들어오는 경우가 있을까요?

SightStudio replied:
이거 초기 설계엔 들어올 수도 있다고 생각해서 납뒀었네요.. 
수정하도록 하겠습니다.

jinyoungchoi95 commented on src/main/java/orm/StatefulPersistenceContext.java:
```diff
@@ -62,6 +83,47 @@ public <T> Object getDatabaseSnapshot(EntityIdHolder<T> idHolder, EntityPersiste
+            return entryContext.addEntry(entityKey, status);
+        }
+
+        // 상태가 이전과 동일하면 add 하지 않음
+        if (entry.getStatus() == status) {
+            return entry;
+        }
```
(반영하지 않으셔도 됩니다.) `entryContext.addEntry`가 map의 put과 같은 단순한 동작을 지원한다면 해당 if문은 없어도 동일하게 동작할 수 있으므로 메소드를 읽는 컨텍스트를 늘리는 행위가 될 수 있다고 생각합니다 :)

jinyoungchoi95 reviewed (changes requested):
안녕하세요 동민님 😄

몇가지 코멘트 남겨두었는데 확인부탁드려요 :)

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

SightStudio committed:
refactor. 리뷰반영

1. TrackableEntityEntryContext.java 테스트 패키지로 위치변경

2. Status.java 주석 제거

3. SimpleEntityEntry.java ID대신 EntityKey를 가지고 있도록

4. SessionImpl.java에서 불필요한 엔트리 추가행위 제거

SightStudio committed:
refactor. 리뷰반영, StatefulPersistenceContext.addEntity(T entity) 에서 널체크 제거

jinyoungchoi95 commented on src/main/java/orm/life_cycle/SimpleEntityEntry.java:
```diff
@@ -0,0 +1,35 @@
+
+import orm.EntityKey;
+
+public class SimpleEntityEntry implements EntityEntry {
+
+    private final EntityKey entityKey;
```
엄밀히따지면 context에서 이미 map으로 `EntityKey`를 들고 있기 때문에 현재는 `EntityEntry`가 key를 들고 있을 필요는 없어보여요.

jinyoungchoi95 reviewed (approved):
안녕하세요 동민님 😄

미션 수고하셨습니다 👍
다음 단계도 화이팅하세요!

감사합니다 🙇‍♂️
